LDAP: Refactor and add ldap_passwd module

KellerFuchs opened the pull request:
##### SUMMARY
- [x] Refactor `ldap_attr` and `ldap_entry`, pulling shared code, documentation and specs in `__init__`.
- [x] Add a `ldap_passwd` module.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
notification/ldap

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 17 2017, 18:50:44) [GCC 7.2.0]
```

##### Additional information

I needed such a module while working on [LDAP-related automation](https://github.com/hashbang/admin-tools/pull/112).

KellerFuchs committed:
modules/net_tools/ldap: Refactor shared options

KellerFuchs committed:
modules/net_tools/ldap: Refactor shared code

KellerFuchs committed:
modules/net_tools/ldap: Add ldap_passwd module

Akasurde commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,64 @@
+COMMON_SPEC={
```
I would prefer `ldap_argument_spec`

Akasurde commented on lib/ansible/modules/net_tools/ldap/ldap_attr.py:
```diff
@@ -75,6 +71,18 @@
     default: 'no'
     description:
       - If true, we'll use the START_TLS LDAP extension.
+  validate_certs:
```
Any reason for this move?

Akasurde commented on lib/ansible/modules/net_tools/ldap/ldap_attr.py:
```diff
@@ -186,40 +186,24 @@
-    HAS_LDAP = False
-
 from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ldap import *
```
Please import specific things rather than everything.

Akasurde commented on lib/ansible/modules/net_tools/ldap/ldap_entry.py:
```diff
@@ -149,31 +149,24 @@
 import traceback
 
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ldap import *
```
Please import specific things rather than everything.

Akasurde commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,150 @@
+    required: true
+    default: null
+    description:
+      - The (plaintext) password to be set for I(dn).
```
You can create documentation fragment for common parameters. You can see pre-existing documentation fragment for [ipa](https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/ipa.py).

KellerFuchs replied:
Thanks for the pointer; I tried factoring out the documentation of the common parameters in an earlier version of the PR, but didn't know this should be done through `module_docs_fragments`.

Akasurde commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2017, Keller Fuchs <kellerfuchs@hashbang.sh>
```
```# Copyright: (c) 2017,```

Akasurde commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,64 @@
+        # Establish connection
+        self.connection = self._connect_to_ldap()
+
+    def _connect_to_ldap(self):
```
module_utils code is BSD licensed and modules are GNU GPLv3 licensed. I think we need to get sign-off from everyone who contributed to this code snippet before moving it to module_utils. You can refer this [comment](https://github.com/ansible/ansible/pull/26613#issuecomment-318097595) for process.

KellerFuchs commented:
@psagers @jtyr During that refactoring I moved code from your `ldap_*` modules to a shared helper in `module_utils`. However, this involves a change of license (from GPLv3 to BSD) for the code that moved to `modules_utils`.
I would like to have your permission to relicense the code that was moved to `ldap.py`.

psagers commented:
Couldn't possibly care less. I pushed a change to the original repository, re-releasing the code under CC0. I'll leave the repo up a little longer.

KellerFuchs commented:
@psagers OK, thanks.

KellerFuchs commented:
ready_for_review

KellerFuchs commented:
@Akasurde I think I handled all your comments (pending OK from @jtyr for relicensing).

KellerFuchs committed:
modules/net_tools/ldap/ldap_passwd: More robust change check

In some deployments, using compare_s results in spurious “changed” results,
while bind is more reliable.  The downside is that it results in an extra
connection, and the code it more involved.

KellerFuchs commented:
@jtyr Ping?  This PR is blocked on your OK for relicensing some of the code.

jtyr commented:
I'm fine with change of the license but would prefer to change of `modules_utils` to GPLv3 instead of changing the modules to BSD ;o)

KellerFuchs commented:
@jtyr Am I right in understanding this as an OK (albeit a facetious one) ?

If so, we only need @Akasurde to confirm that the changes they requested were done.

KellerFuchs commented:
@Akasurde Ping?  This should be good to merge.

Akasurde commented:
@KellerFuchs Could you please address ansibot reported issues ?

KellerFuchs commented:
@Akasurde I was currently looking at that, and it should be fixed (we will see once the CI run); it took a while because the error message was somewhat confusing: I think the actual error was that there was both `type: bool` and `choices: [yes, no]` set in the documentation fragment.

That existed prior to the PR, and was fixed by @dagwieers in cdd21e2170b9a47d84624c8c2f792a66f2b0fe57, but I accidentally reintroduced it during the rebase.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+
+    ldap = LdapPasswd(module)
+
+    try:
```
I would rather remove this exception here and put it around the `ldap.initialize()` and `self.connection.passwd_s()` above with a more specific error message.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+        # Shortcuts
+        self.passwd = self.module.params['passwd']
+
+    def passwd_c(self):
```
I would name the method `passwd_check`.

KellerFuchs replied:
This followed the conventions from the LDAP library, but we can change that.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+        finally:
+            u_con.unbind()
+
+    def passwd_s(self):
```
I would name the method `passwd_set`.

KellerFuchs replied:
Same as above.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+
+def main():
+    module = AnsibleModule(
+        argument_spec=gen_specs(passwd=dict(type='str', no_log=True)),
```
You can remove `type='str'` as that's the default type.

jtyr commented on lib/ansible/utils/module_docs_fragments/ldap.py:
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2017 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
```
This file consists of texts from previously copyrighted file. You should respect the original copyright.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+            try:
+                u_con.start_tls_s()
+            except ldap.LDAPError as e:
+                self.module.fail_json(msg="Cannot start TLS.", details=to_native(e),
```
Break the line after the `(` to unify the style across the file.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+        else:
+            module.exit_json(changed=ldap.passwd_s())
+    except Exception as e:
+        module.fail_json(msg="Passwd action failed.",
```
Break the line after the `(` to unify the style across the file.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
I would format this like this:

```python
ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}
```

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+    dn: "{{ item.key }}"
+    passwd: "{{ item.value }}"
+  with_dict:
+    alice: "alice123123"
```
You don't have to quote this string.

KellerFuchs replied:
I did so for consistency, but I can remove the quoting.

jtyr commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,63 @@
+import traceback
```
This file is missing copyright.

KellerFuchs replied:
For some reason, Github is not dismissing the comment, but I added the copyright info.

jtyr commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,63 @@
+            try:
+                connection.start_tls_s()
+            except ldap.LDAPError as e:
+                self.module.fail_json(msg="Cannot start TLS.", details=to_native(e),
```
Break the line after the `(` to unify the style across the file.

KellerFuchs committed:
ldap_passwd: Rename methods passwd_[cs]

KellerFuchs committed:
Update copyright info for ldap modules

KellerFuchs committed:
ldap_passwd: Remove unecessary type=str

KellerFuchs committed:
ldap: Factor-out failure cases

KellerFuchs committed:
ldap_passwd: Provide more precise error messages

KellerFuchs committed:
ldap_passwd: Irrelevant syntax changes

KellerFuchs commented:
@jtyr Comments handled, please ACK.

jtyr commented on lib/ansible/utils/module_docs_fragments/ldap.py:
```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2017-2018 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
```
The right order should be:

```
# Copyright 2017-2018 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
# Copyright 2016 Jiri Tyr <jiri.tyr@gmail.com>
# Copyright 2016 Peter Sagerson <psagers@ignorare.net>
```

KellerFuchs replied:
This was literally copy-pasted from the original file.

jtyr replied:
The original modules have this:

```yaml
# (c) 2016, Peter Sagerson <psagers@ignorare.net>
# (c) 2016, Jiri Tyr <jiri.tyr@gmail.com>
```

Peter is the original author, I'm the person who polished and published his module. You are another contributor, so you should be at the end of the list.

KellerFuchs replied:
Thanks for clarifying, because it was utterly unclear to me that those lines should be in chronological order, rather than reverse chronological + alphabetical order.

KellerFuchs replied:
Eeeeeh, this is literally the opposite of [what you previously said](https://github.com/ansible/ansible/pull/33040#discussion_r182894470)

RyanSquared replied:
> You are another contributor, so you should be at the end of the list.

And yet for your comment six days ago, @KellerFuchs was on top?

jtyr commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,86 @@
+    import ldap
+    import ldap.sasl
+    HAS_LDAP = True
+
```
I would move this blank line above the `HAS_LDAP`.

KellerFuchs replied:
What's the rationale for it?

RyanSquared replied:
A lot of Python code typically has a blank line after imports.

jtyr replied:
Exactly.  The new line should be right after the import and not after the global variable.

jtyr commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,86 @@
+        self.connection = self._connect_to_ldap()
+
+    def fail(self, msg, exn):
+        self.module.fail_json(msg=msg,
```
Please wrap the line after the `(`.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ldap import LdapGeneric, gen_specs, HAS_LDAP
```
I would keep the `HAS_LDAP` local to this package. Just set it to `True` after you import it bellow.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+
+    def passwd_check(self):
+        try:
+            u_con = ldap.initialize(self.server_uri)
```
What does mean the `u` in  `u_con`? What about to call it just `con`?

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_passwd.py:
```diff
@@ -0,0 +1,149 @@
+        except ldap.LDAPError as e:
+            self.fail("Unable to set password", e)
+
+        # We successfully changed the password  \o/
```
I would remove the emoji here ;o)

robbyoconnor replied:
awww no fun whatsoever!

KellerFuchs committed:
ldap_passwd: Rename u_con to tmp_con

KellerFuchs committed:
ldap_passwd: Keep HAS_LDAP local

KellerFuchs committed:
ldap: Reorder copyright lines

KellerFuchs committed:
module_utils/ldap: Reindent

KellerFuchs commented:
@jtyr FYI, if those are a syntax convention that Ansible actually uses and enforces, there should be a linter running as part of the CI.  It's not a good use or your or my time to go back and forth about what essentially amounts to indentation changes and reshuffling copyright lines.

Moreover, it would make the requests feel far less arbitrary, esp. when they concern things that were literally copy-pasted from Ansible's codebase...

KellerFuchs committed:
ldap_passwd: Fun is forbidden

jtyr commented:
I'm sorry for being so nit picking but your change is based on my work so it's in my interest to maintain the same quality of the code.

KellerFuchs committed:
fixup! module_utils/ldap: Reindent

KellerFuchs commented:
@jtyr My issue is not about you being nitpicky.

My issue is that you ignored this PR for months, and now that you deign pay attention to it, I am getting seemingly-arbitrary requests to change things, and when I comply with those, I get a contradictory request right afterwards.  Case in point:
- you provide a [specific order](https://github.com/ansible/ansible/pull/33040#discussion_r182894470) you want for the copyright lines;
- I comply with the request in 6549474da59428350cd7005ea4b830d7c99a4e0d, uniformizing that on all concerned files (incl. the ones I did not create);
- not even a week later [you request the exact opposite](https://github.com/ansible/ansible/pull/33040#discussion_r184024375).

If you truly feel that the code quality is improved by the changes, feel free to push commits doing exactly that to this branch; Github has an “allow edits from maintainers” feature for a reason.
In any case, I am entirely done wasting my time complying with contradictory, pointless changes; please let me know if you have actually-relevant review comments...

KellerFuchs committed:
fixup! Update copyright info for ldap modules

Akasurde committed:
LDAP doc update

* Resolved all copyright related issues
* Resolved self.fail calls
* Update documentation

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde commented:
@KellerFuchs @jtyr I updated PR with required changes. Let me know if we agree on changes, so that we can merge changes.

jtyr commented on lib/ansible/utils/module_docs_fragments/ldap.py:
```diff
@@ -0,0 +1,40 @@
+options:
+  bind_dn:
+    description:
+    - A DN to bind with. If this is omitted, we'll try a SASL bind with the EXTERNAL mechanism.
```
You should indent all the lists in this file by two more spaces to keep it the same like in all other files.

jtyr commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,76 @@
+# Copyright: (c) 2016, Peter Sagerson <psagers@ignorare.net>
+# Copyright: (c) 2016, Jiri Tyr <jiri.tyr@gmail.com>
+# Copyright: (c) 2017-2018 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
This should look like this:

```yaml
# -*- coding: utf-8 -*-

# Copyright: (c) 2016, Peter Sagerson <psagers@ignorare.net>
# Copyright: (c) 2016, Jiri Tyr <jiri.tyr@gmail.com>
# Copyright: (c) 2017-2018 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
#
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

The same in all other files.

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_attr.py:
```diff
@@ -1,21 +1,21 @@
 
 
-DOCUMENTATION = """
+DOCUMENTATION = r"""
```
Please remove the `r` here as there is no reason to use byte string for any of the string blocks. The same in all other files.

webknjaz replied:
correction: It's not a byte string, it's a raw-string literal

jtyr commented on lib/ansible/modules/net_tools/ldap/ldap_attr.py:
```diff
@@ -178,40 +149,30 @@
     import ldap
-    import ldap.sasl
-
-    HAS_LDAP = True
```
You should keep the `HAS_LDAP = True` here and remove the import for it. The same in all other files.

jtyr commented on lib/ansible/module_utils/ldap.py:
```diff
@@ -0,0 +1,76 @@
+        # Establish connection
+        self.connection = self._connect_to_ldap()
+
+    def fail(self, msg, exn):
```
I like this simplification but I believe that it should done somewhere in the core (suggestion for another PR) and not on per-module bases.

gundalow replied:
Perhaps this can be done in a future PR?
That way there can be focused discussion about one topic in a single PR?

gundalow replied:
ie this PR is fine, we can look at moving into Core at a later point

Akasurde replied:
Agreed.

gundalow commented:
@jtyr Thanks for the review comments so far, can you confirm you are happy with the structure and the only remaining technical issues are

1. `HAS_LDAP` handeling
2. `fail(self, msg, exn):`, - we will keep what we have in this PR and look at moving this out in a follow up PR

jtyr commented:
In general I'm happy with this PR. After the few details are ironed out, I'm happy to approve it.

Akasurde committed:
Include code review comments

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde commented:
@KellerFuchs Thanks for the contribution. @jtyr @robbyoconnor @gundalow @webknjaz @RyanSquared Thanks for the reviews and comments.

KellerFuchs commented:
@Akasurde You are welcome; thanks for the review and fixups.
